# Random Forest Classifier

In [10]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report, confusion_matrix

In [11]:
iris = sns.load_dataset('iris')

In [12]:
#subset the data 
x = iris.drop('species', axis=1)
y = iris['species']

In [13]:
#train test split the data 
x_train, x_test, y_train, y_test = train_test_split(x,y, test_size=0.3, random_state=76)

### When you are doing this normally, you can skip making the single tree; the decision tree is mostly here for comparison. That means that you can run this code immediately after splitting your data into training and testing sets.

#### Initial Random Forest Model
And at last you are ready to kick it into high gear by creating your random forest model. You'll use the function RandomForestClassifer(), with the arguments n_estimators= to specify how many decision trees you want the random forest to stem from, and of course random_state= just to follow along with this content:

In [22]:
forest = RandomForestClassifier(n_estimators=500, random_state=76)
forest.fit(x_train, y_train)

RandomForestClassifier(n_estimators=500, random_state=76)

#### RandomForestClassifier?   adding question mark after a function gives possible arguments and info

#### This is exactly like when you set up a single tree, except now you specify the number of trees you want to make with n_estimators=, which you have set to 500. This means you will be testing your data with 500 decision trees. If you have a very large dataset, you may want to set n_estimators= smaller to decrease the time it takes to run the RandomForestClassifier() function. However, in general, the higher your n_estimators=, the more accurate your model will be. You'll then fit() to the data. Here is the output:

In [20]:
RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=500, n_jobs=None,
            oob_score=False, random_state=76, verbose=0, warm_start=False)

RandomForestClassifier(n_estimators=500, random_state=76)

In [23]:
pwd

'C:\\Users\\Marcy\\Documents\\Entity Coursework\\DS106 Machine Learning\\Random Forests'

#### Evaluate Model Fit
The final step is to create your prediction set and print a report! Again, you'll make use of both the confusion matrix and the classification report.

In [21]:
forestPredictions = forest.predict(x_test)
print(confusion_matrix(y_test, forestPredictions))
print(classification_report(y_test, forestPredictions))

[[19  0  0]
 [ 0 11  2]
 [ 0  0 13]]
              precision    recall  f1-score   support

      setosa       1.00      1.00      1.00        19
  versicolor       1.00      0.85      0.92        13
   virginica       0.87      1.00      0.93        13

    accuracy                           0.96        45
   macro avg       0.96      0.95      0.95        45
weighted avg       0.96      0.96      0.96        45



[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done 500 out of 500 | elapsed:    0.0s finished


#### Looks like the random forest did better than the decision tree, which is to be expected. All setosa and virginica irises were correctly classified, and only two of the versicolor irises were misclassified as virginica.

Now, looking at the classification report, again the random forest model did a better job:

In [25]:
            precision    recall  f1-score   support

      setosa       1.00      1.00      1.00        19
  versicolor       1.00      0.85      0.92        13
   virginica       0.87      1.00      0.93        13

   micro avg       0.96      0.96      0.96        45
   macro avg       0.96      0.95      0.95        45
weighted avg       0.96      0.96      0.96        45

IndentationError: unindent does not match any outer indentation level (<tokenize>, line 4)

#### That is it. The new model is 96% accurate, and in general it will become more and more accurate the larger your dataset is. There is 100% accuracy for both setosa and versicolor irises, but only 87% accuracy for virginica. Go forth and explore the Random Forest!